### PR TITLE
Configurable value replacement on match failure for RegexExtractionFn

### PR DESCRIPTION
**Pull request by jon-wei**
Addresses issue #2064 

This adds 3 new parameters to the RegexExtractionFn:

replaceMissingValues: Boolean, if a regex match fails, replace the dimval with a user specified value instead of returning the orignal dimval, default is false

replaceMissingValuesWith: String, replace dimval with this value when regex match fails if replaceMissingValues == true

injective: Boolean, currently unused, but same usage as "injective" property in the JavaScriptExtractionFn and LookupExtractionFn

### CODE REVIEW

File: processing/src/main/java/io/druid/query/extraction/RegexDimExtractionFn.java [hunk 1]
```diff
@@ -34,28 +35,59 @@
...
+  private final String replaceMissingValuesWith;
 
   @JsonCreator
   public RegexDimExtractionFn(
-      @JsonProperty("expr") String expr
-  )
+      @JsonProperty("expr") String expr,
+      @JsonProperty("replaceMissingValues") Boolean replaceMissingValues,
+      @JsonProperty("replaceMissingValuesWith") String replaceMissingValuesWith,
+      @JsonProperty("injective") Boolean injective
```
Comment (b-slim) [inline processing/src/main/java/io/druid/query/extraction/RegexDimExtractionFn.java]: i don't see why we need this property ? it  relevant for function lookup but i don't think it is useful for regex. If you have any use case please let me know.

File: processing/src/test/java/io/druid/query/extraction/RegexDimExtractionFnTest.java [hunk 2]
```diff
@@ -117,13 +117,54 @@ public void testNullAndEmpty()
...
+    ExtractionFn extractionFn = new RegexDimExtractionFn(regex, true, "foobar", false);
+    Set<String> extracted = Sets.newHashSet();
+
+    for (String testString : testStrings) {
+      extracted.add(extractionFn.apply(testString));
+    }
+
+    Assert.assertEquals(4, extracted.size());
+    Assert.assertTrue(extracted.contains("apple"));
+    Assert.assertTrue(extracted.contains("awesome"));
```
Comment (b-slim) [inline processing/src/test/java/io/druid/query/extraction/RegexDimExtractionFnTest.java]: arrays comparison will be better thought!

File: processing/src/test/java/io/druid/query/extraction/RegexDimExtractionFnTest.java [hunk 3]
```diff
@@ -117,13 +117,54 @@ public void testNullAndEmpty()
...
+    Assert.assertTrue(extracted.contains("asylum"));
+    Assert.assertTrue(extracted.contains("foobar"));
+
+    byte[] cacheKey = extractionFn.getCacheKey();
+    Assert.assertEquals(17, cacheKey.length);
```
Comment (b-slim) [inline processing/src/test/java/io/druid/query/extraction/RegexDimExtractionFnTest.java]: usually we compare the byte arrays one case when equals and another when not

**Comment by jon-wei** [comment]
@b-slim can you make your comments on the newer PR

**Pull request closed without merging**